### PR TITLE
Add virtualization to the ComboBox

### LINKED ISSUE #3056: ComboBox Performance degrades with many items
Willing to submit PR: Yes

When using the ComboBox with 500+ items, the performance of the callout degrades significantly because there is no virtualization of the options list.

The options list could use the Fabric List component for virtualization, which should improve performance.

### PR DESCRIPTION
**Pull request by abettadapur**
#### Pull request checklist

@cschleiden  @jspurlin 

- [x] Addresses an existing issue: Fixes #3056 
- [x] Include a change request file using `$ npm run change`

#### Description of changes
* Adds a new prop `virtualized` to ComboBox
* If `virtualized` is true, render the ComboBox options in a Fabric List component
* If `virtualized` is true, use the List scrollToIndex function when trying to scroll to the selected element
* Added an example with 1000 items

#### Focus areas to test
* Test scrolling to elements and scroll performance

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Basic.Example.tsx [hunk 1]
```diff
@@ -63,6 +75,20 @@ export class ComboBoxBasicExample extends React.Component<any, any> {
           onBlur={ () => console.log('onBlur called') }
         // tslint:enable:jsx-no-lambda
         />
+        <ComboBox
+          defaultSelectedKey='C'
```
Comment (abettadapur) [inline packages/office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Basic.Example.tsx]: Remove this example

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.Props.ts [hunk 2]
```diff
@@ -95,6 +95,12 @@ export interface IComboBoxProps extends ISelectableDroppableTextProps<IComboBox>
    */
   scrollSelectedToTop?: boolean;
 
+  /**
+   * Should the combo box list be virtualized? Useful for combo boxes that contain a large number of items
+   * @default false
+   */
+  virtualized?: boolean;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.Props.ts]: This is great! But it might be worth splitting into `VirtualizedComboBox`. That way we keep the bundle size of ComboBox minimal, and only pull List in when needed. What do you think?

Adding @christiango and @cschleiden here as well.

Comment (abettadapur) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.Props.ts]: That sounds reasonable to me, VirtualizedComboBox can extend from ComboBox and I can make the scrollTo function protected

**Comment by abettadapur** [comment]
This now depends on #3059

File: packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import * as React from 'react';
+import { autobind, BaseComponent, customizable } from '../../Utilities';
+import { ComboBox } from './ComboBox';
+import { IComboBoxProps } from './ComboBox.Props';
+import { List } from "../../List";
+import { SelectableOptionMenuItemType, ISelectableOption } from '../../utilities/selectableOption/SelectableOption.Props';
+
+export class VirtualizedComboBox extends BaseComponent<IComboBoxProps, {}> {
+  // The list element, if the combobox is virtualized
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx]: nit, should be always virtualized?

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 4]
```diff
@@ -784,23 +788,22 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
...
+    // Render using regular mapping
     return (
       <div
         id={ id + '-list' }
         className={ this._classNames.optionsContainer }
         aria-labelledby={ id + '-label' }
         role='listbox'
       >
-        { this.state.currentOptions.map((item, index) => onRenderItem({ ...item, index } as ISelectableOption, this._onRenderItem)) }
+        { props.options.map((item) => onRenderItem(item, this._onRenderItem)) }
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Is this change needed anymore?

Comment (abettadapur) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Leaked from the other change, thanks. Will remove

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 5]
```diff
@@ -3,6 +3,7 @@ import { IComboBoxOption, IComboBoxProps } from './ComboBox.Props';
 import { DirectionalHint } from '../../common/DirectionalHint';
 import { Callout } from '../../Callout';
 import { Label } from '../../Label';
+import { List } from '../../List';
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: remove this and the field below

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 6]
```diff
@@ -784,23 +788,22 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
     } = this.props;
 
     let id = this._id;
-
+    // Render using regular mapping
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: remove the comment?

File: packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        role='listbox'
+        items={ props.options }
+        onRenderCell={ onRenderItem != null ? (item: ISelectableOption) => onRenderItem(item) : () => null }
+      />
+    );
+  }
+
+  protected _onScrollToItem(itemIndex: number): void {
+    // We are using the List component, call scrollToIndex
+    this._list.scrollToIndex(itemIndex);
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx]: Does this blow up if itemIndex is null/undefined or is not a valid index (e.g. -1)?

Comment (abettadapur) [reply packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx]: I don't believe so, the relevant check is `itemIndex <= index && (itemIndex + itemsPerPage) > index;`, which will always fail if null or negative. It ends up just being a no-op

Comment (abettadapur) [reply packages/office-ui-fabric-react/src/components/ComboBox/VirtualizedComboBox.tsx]: See https://github.com/OfficeDev/office-ui-fabric-react/blob/246e162e1c5573cc597d33e6c1726dbc8c6e457c/packages/office-ui-fabric-react/src/components/List/List.tsx#L179 for the implementation
